### ISSUE TITLE
Memory & errors in workers

### ISSUE THREAD

**Issue by neowutran** [actix/actix-web#763]
Hi, 

Some code / errors that lead to high memory usage ( / memory leak ). 
(Tested with latest commit on master branch)

main.rs
```rust
use actix_web::{web, App, HttpServer};
use std::collections::HashMap;

fn no_params() -> &'static str {
  let map: HashMap<String,String> = HashMap::new();
  let _error_before = map.get("DontExist").unwrap();
    "Hello world!\r\n"
}

fn main() -> std::io::Result<()> {
  HttpServer::new(|| {
    App::new()
    .service(web::resource("/test").route(web::post().to(no_params)))
  })
   .bind("127.0.0.1:80")?
    .run()
}
```

Cargo.toml
```
[dependencies]
actix-web = {git="https://github.com/actix/actix-web"}
```

client.sh
```bash
#!/bin/bash
while true; do curl -X POST http://127.0.0.1/test; done
```

stdout of the binary
```
...
thread 'actix-rt:worker:22573' panicked at 'called `Option::unwrap()` on a `None` value', src/libcore/option.rs:345:21
Panic in Arbiter thread, shutting down system.
thread 'actix-rt:worker:22574' panicked at 'called `Option::unwrap()` on a `None` value', src/libcore/option.rs:345:21
Panic in Arbiter thread, shutting down system.
thread 'actix-rt:worker:22575' panicked at 'called `Option::unwrap()` on a `None` value', src/libcore/option.rs:345:21
Panic in Arbiter thread, shutting down system.
thread 'actix-rt:worker:22576' panicked at 'called `Option::unwrap()` on a `None` value', src/libcore/option.rs:345:21
...
```

Memory usage increase over time. After ~1min, it is at ~100mo. 
Even after client.sh have been killed, memory usage never goes down

**Comment by fafhrd91** [comment]
i dont think we can control what happen with memory after panic. actix-web restarts thread but panicked thread can easily leak memory and i dont see any way how we can reclaim it.

**Comment by fafhrd91** [comment]
i'd suggest to use `panic = abort`

**Comment by mattlockyer** [comment]
Wanted to provide some context to @fafhrd91 comments above.
Using `panic = abort` will propagate to the main thread and crash your server, not ideal.

**TLDR**
The best strategy is to ensure you're handling all `Option<T>` or `Result<T, E>` errors and responding appropriately to the client.
Helpful: https://doc.rust-lang.org/edition-guide/rust-2018/error-handling-and-panics/the-question-mark-operator-for-easier-error-handling.html

Some more details about rust memory safety here:
https://www.reddit.com/r/rust/comments/9q3jqn/how_is_rust_safe_when_panics_can_happen_out_of/

Detecting panic! at compile time:
https://crates.io/crates/no-panic will cause a compiler error if code is found to panic. You could put this on your request handlers. I haven't found the error messages to be particularly helpful though.

Reading deeper into the Reddit thread...
For the bold there's this: https://doc.rust-lang.org/std/panic/fn.catch_unwind.html
Which signals there may be a way to catch panic-ing worker threads in the Actix App / Sys level and free their memory by implementing `Drop`. I'm not that advanced with Rust yet. Is this even a good idea?

Conclusion (same as TLDR)
The best course of action is to simply handle all errors in request handlers using ? or `match`.
